File: C/pa2/backup/List.c
//Samer Baslan
//sbaslan
//CMPS 101
//10-17-2017
//List.c

#include<stdio.h>
#include<stdlib.h>
#include "List.h"


// private NodeObj type
typedef struct NodeObj{
   int data;
   struct NodeObj* next;
   struct NodeObj* prev;
} NodeObj;

// private Node type
typedef NodeObj* Node;
	
// private ListObj type
typedef struct ListObj{
   Node front;
   Node back;
   Node cursor;
   int cursorindex;
   int length;
} ListObj;

// Constructors and Destructors ----------------------------------------------------

// newNode()
// Returns reference to new Node object. Initializes next and data fields.
// Private.
Node newNode(int data){
   Node N = malloc(sizeof(NodeObj));
   N->data = data;
   N->next = NULL;
   N->prev = NULL;
   return(N);
}

// freeNode()
// Frees heap memory pointed to by *pN, sets *pN to NULL.
// Private.

void freeNode(Node* pN){
   if(pN != NULL && *pN != NULL){
      free(*pN);
      *pN = NULL;
   }
}

// newList()
// Returns reference to new empty Queue object.
List newList(void){
   List L;
   L = malloc(sizeof(ListObj));
   L->front = L->back = L->cursor = NULL;
   L->length = 0;
   L->cursorindex = -1;
   return(L);
}

// freeList()
// Frees all heap memory associated with Queue *pL, and sets *pL to NULL.S
void freeList(List* pL){
   if(pL != NULL && *pL != NULL){
      Node N = (*pL)->front;
      while(N != NULL){
         Node temp = N;
         N = N->next;
         free(temp);         // use temp because wont be able to point to N after it is freed
      }
      free(*pL);
      *pL = NULL;
   }
}
      
// Access functions------------------------------------------------------------
// length()
// Returns the number of elements in this List.
int length(List L){
   if(L==NULL){
      printf("List Error: calling length() on NULL List reference\n");
      exit(1);
   }
   return L->length;
}

// index()
// If cursor is defined, returns the index of the cursor element.
// other wisereturns -1.
int index(List L){
   if(L==NULL){
      printf("List Error: calling index() on NULL List reference\n");
      exit(1);
   }
   return L->cursorindex;
}

// front()
// Returns front element.
// Pre: length()>0
int front(List L){
   if(L==NULL){
      printf("List Error: calling front() on NULL List reference\n");
      exit(1);
   }
   return L->front->data;
}

// back()
// Returns back element.
// Pre: length()>0 
int back(List L){
   if(L==NULL){
      printf("List Error: calling back() on NULL List reference\n");
      exit(1);
   }
   return L->back->data;
}

// get()
// Returns cursor element.
// Pre: length()>0, index()>=0
int get(List L){
   if(L==NULL){
      printf("List Error: calling get() on NULL List reference\n");
      exit(1);
   }
   if(L->cursorindex < 0){
      printf("List Error: calling get() on undefined cursor");
   }
   return L->cursor->data;
}
   
// equals()
// Returns true if this List and L are the same integer
// sequence. The cursor is ignored in both lists.
int equals(List A, List B){
   if(A == NULL || B == NULL){
      printf("List Error: calling equals() on empty List reference\n");      exit(1);
   } 
   if(A->length == B->length){
      Node tempfront = B->front;
      Node tempfront2 = A->front;
      while(tempfront->next != NULL && tempfront2->next != NULL){
         if(tempfront->data == tempfront2->data){
            tempfront = tempfront->next;
            tempfront2 = tempfront2->next;
         }
         else{
            return 0;
         }
      }
      return 1;
   }
   return 0;
}
      
// Manipulation procedures ----------------------------------------------------

// clear()
// Resets this list to its original empty state.
void clear(List L){
   if(L == NULL){
      printf("List Error: calling clear() on NULL List reference\n");
      exit(1);
   }
/*  Node N = NULL;
   for(N = L->front; N != NULL; N = N->next){ //Free before clearing
      Node temp = N;
      freeNode(&temp);
   }*/
   
   freeNode(&L->front);
   freeNode(&L->back);
   freeNode(&L->cursor);
   L->front = L->back = L->cursor = NULL;
   L->length = 0;
   L->cursorindex = -1;
}
// moveFront()
// If List is non-empty, places the cursor under the front element,
// otherwise does nothing.
void moveFront(List L){
   if(L == NULL){
      printf("List Error: calling moveFront() on NULL List reference\n");
      exit(1);
   }
   if(L->length > 0){
      L->cursor = L->front;
      L->cursorindex = 0;
   }
}
// moveBack()
// If List is non-empty, places the cursor under the back element,
// otherwise does nothing.
void moveBack(List L){
   if(L == NULL){
      printf("List Error: calling moveBack() on NULL List reference\n");
      exit(1);
   }
   if(L->length > 0){
      L->cursor = L->back;
      L->cursorindex = L->length - 1;
   }
}

// movePrev()
// If cursor is defined and not at front, moves cursor one step towards
// front of this List, if cursor is defined and at front, cursor becomes
// undefined, if cursor is defined does nothing.
void movePrev(List L){
   if(L == NULL){
      printf("List Error: calling movePrev() on NULL List reference\n");
      exit(1);
   }
   if(L->cursor != NULL && L->cursorindex != 0){
      L->cursor = L->cursor->prev;
      L->cursorindex = L->cursorindex - 1;
   }
   else if(L->cursor != NULL && L->cursorindex <= 0){
      L->cursor = NULL;
      L->cursorindex = -1;
   }
   else{ //do nothing
   }
}
   
      
// moveNext()
// If cursor is defined and not at back, moves cursor one step toward
// back of this List, if cursor is defined and at back, cursor beomes
// undefined, if cursor is undefined does nothing.
void moveNext(List L){
   if(L == NULL){
      printf("List Error: calling moveNext() on NULL List reference\n");
      exit(1);
   }
   if(L->cursor != NULL && L->cursorindex != L->length-1){
      L->cursor = L->cursor->next;
      L->cursorindex = L->cursorindex + 1;
   } 
   else if(L->cursor != NULL && L->cursorindex >= L->length-1){
      L->cursor = NULL;
      L->cursorindex = -1;
   }
   else{ // do nothing
   }
} 

// prepend()
// Insert new element into this List. If List is non-empty,
// insertion takes place before front element.
void prepend(List L, int data){
   if(L == NULL){
      printf("List Error: calling prepend() on NULL List reference\n");
      exit(1);
   }
  
   Node N = newNode(data);
   if(L->length == 0){
      L->front = L->back = N;
   }
   else{
      L->front->prev = N;
      N->next = L->front;
      L->front = N;
   }
   L->length = L->length + 1;
}

// append()
// Insert new element into this List. If List is non-empty,
// insertion takes place after back element.
void append(List L, int data){
   if(L == NULL){
      printf("List Error: calling append() on NULL List reference\n");
      exit(1);
   }
   Node N = newNode(data);
   if(L->length == 0){
      L->front = L->back = N;
   }
   else{
      L->back->next = N;
      N->prev = L->back;
      L->back = N;
   }
   L->length = L->length + 1;
}

// insertBefore()
// Insert new element before cursor.
// Pre: length()>0, index()>=0
void insertBefore(List L, int data){
   if(L == NULL){
      printf("List Error: calling insertBefore() on NULL List reference\n");
      exit(1);
   }
   if(L->length <= 0){
      printf("List Error: calling insertBefore() called on empty List");
      exit(1);
   }
   if(L->cursorindex < 0){
      printf("List Error: calling insertBefore() called on undefined cursor");
      exit(1);
   }
   Node N = newNode(data);
   if(L->cursor->prev == NULL){
      L->front = N;
      L->cursor->prev = N;
      N->next = L->cursor;
   }
   else{
      N->prev = L->cursor->prev;
      L->cursor->prev->next = N;
      N->next = L->cursor;
   }
   L->length = L->length + 1;
   L->cursorindex = L->cursorindex + 1;
}

// insertAfter()
// Insert new element after cursor.
// Pre: length()>0, index()>=0
void insertAfter(List L, int data){
   if(L == NULL){
      printf("List Error: calling insertAfter() on NULL List reference\n");
      exit(1);
   }
   if(L->length <= 0){
      printf("List Error: calling insertAfter() on empty list\n");
      exit(1);
   }
   if(L->cursorindex < 0){
      printf("List Error: calling insertAfter() on undefined cursor\n");
      exit(1);
   }
   Node N = newNode(data);
   if(L->cursor->next == NULL){
      L->back = N;
      L->cursor->next = N;
      N->prev = L->cursor;
   }
   else{
      L->cursor->next->prev = N;
      N->next = L->cursor->next;
      N->prev = L->cursor;
      L->cursor->next = N;
   }    
   L->length = L->length + 1;
   // dont need to increment cursor index because the element is inserted
   // after the cursor
}

// deleteFront()
// Deletes the front element.
// Pre: length()>0
void deleteFront(List L){
   if(L == NULL){
      printf("List Error: calling deleteFront() on NULL List reference\n");
      exit(1);
   }
   if(L->length <= 0){
      printf("List Error: calling deleteFront() on empty List");
      exit(1);
   }
   L->front = L->front->next;
   L->front->prev = NULL;
   L->length = L->length - 1;
} 
      

// deleteBack()
// Deletes the back element.
// Pre: length()>0
void deleteBack(List L){
   if(L == NULL){
      printf("List Error: calling deleteBack() on NULL List reference\n");
      exit(1);
   }
   if(L->length <= 0){
      printf("List Error: calling deleteBack() on empty List\n");
      exit(1);
   }
   L->back = L->back->prev;
   L->back->next = NULL;
   L->length = L->length - 1;
}
   

// delete()
// Deletes the cursor element, making cursor undefined.
// Pre: length()>0, index()>=0
void delete(List L){
   if(L == NULL){
      printf("List Error: calling delete() on NULL List reference\n");
      exit(1);
   }
   if(L->length <= 0){
      printf("List Error: calling delete() on empty List\n");
      exit(1);
   }
   if(L->cursorindex < 0){
      printf("List Error: calling delete() on undefined cursor\n");
      exit(1);
   }
   if(L->cursor == L->front){
      deleteFront(L);
   }
   else if(L->cursor == L->back){
      deleteBack(L);
   }
   else{ // cursor is in the middle
      L->cursor->next->prev = L->cursor->prev;
      L->cursor->prev->next = L->cursor->next;
      L->length = L->length -1;
      L->cursor = NULL;
      L->cursorindex = -1;
   }
}

// Other Functions ------------------------------------------------------------

// printList()
// Prints a String representation of this List
void printList(FILE* out, List L){
/*   if ((out == NULL) || (L == NULL)){
      printf("List Error: calling printList() on NULL List reference\n");
      exit(1);
   }*/
   Node N = NULL;
   for(N = L->front; N!= NULL; N = N->next){
      fprintf(out, "%d ", N->data);
   }
}

// copyList()
// Returns a new List representing the same integer sequence as this
// List. The cursor in the new list is undefined, regardless of the 
// state of the cursor in this List. This List is unchanged.
List copyList(List L){
   List temp = newList();
   Node N = NULL;
   for(N = L->front; N!= NULL; N = N->next){
      append(temp, N->data);
   }
   return temp;
}

File: Java/pa1/backup/List.java
//Samer Baslan
//sbaslan
//CMPS101 Fall 2017
//10-06-2017
//List.java


class List{

	private class Node{
		//Fields
		int data;
		Node next;
		Node prev;

		//Constructor
		Node(int data){
			this.data = data;
			next = null;
			prev = null;
		}

		//toString(): overrides Object's toString() method
		public String toString(){
			return String.valueOf(data);
		}

		//equals(): overrides Object's equals() method
		public boolean equals(Object x){
			boolean eq = false;
			Node that;
			if(x instanceof Node){
				that = (Node) x;
				eq = (this.data==that.data);
			}
			return eq;
		}
	}


	//Fields

	private Node front;
	private Node back;
	private Node cursor;
	private int length;
	private int cursorindex;

	//Constructor
	List(){
	front = null;
	back = null;
	cursor = null;
	length = 0;
	cursorindex = -1;
	}

//Access Functions

	// length()
	// Returns the number of elements in the list
	int length(){
		return length;
	}

	// index()
	// If the cursor is undefined, returns the index of the cursor element, otherwise returns -1
	int index(){
		return cursorindex;
	}

	// front()
	// Retruns front element. Pre: length() > 0
	int front(){
		if(this.length() <= 0){
			throw new RuntimeException("List Error: front() called on empty list");
		}
		return front.data;
	}

	// back()
	// Returns back element. Pre: length() > 0
	int back(){
		if(this.length() <= 0){
			throw new RuntimeException("List Error: back() called on empty list");
		}
		return back.data;
	}

	// get()
	// Returns the cursor element. Pre: length() > 0, index() >= 0
	int get(){
		if(this.length() <= 0){
			throw new RuntimeException("List Error: get() called on empty list");
		}
		if(this.index() < 0){
			throw new RuntimeException("List Error: get() called on undefined cursor");
		}
		return cursor.data;
	}

	// equals()
	// Returns true if and only if this List and L are the same integer sequence.
	// The states of the cursors in the two lists are not used in determining equality.
	boolean equals(List L){
		if(L.length == this.length){
			Node tempfront = L.front;
			Node tempfront2 = this.front;
			while(tempfront.next != null && tempfront2.next != null){
				if(tempfront.data == tempfront2.data){
					tempfront = tempfront.next;
					tempfront2 = tempfront2.next;
				}
				else{
					return false;
				}
			}
			return true;
		}
		return false;
	}


//Manipulation procedures


	// clear
	// Resets this List to its original empty state.
	void clear(){
		front = back = cursor = null;
		length = 0;
		cursorindex = -1;
	}


	// moveFront()
	// If List is nonempty, places the cursor under the front element, otherwise does nothing
	void moveFront(){
		if(this.length() > 0){
			cursor =  front;
			cursorindex = 0;
		}
	}


	// moveBack()
	// If List is nonempty, places the cursor under the back element, otherwise does nothing
	void moveBack(){
		if(this.length() > 0){
			cursor = back;
			cursorindex = length - 1;
		}
	}

	// movePrev()
	// If cursor is defined and not at front, moves cursor one step toward
	// front of this List, if cursor is defined and at front, cursor becomes
	// undefined, if cursor is undefined does nothing.
	void movePrev(){
		if(cursor != null && this.index() != 0){
			cursor = cursor.prev;
			cursorindex = cursorindex - 1;
		}
		else if(cursor != null && this.index() <= 0){
			cursor = null;
			cursorindex = -1;
		}
		else{ //do nothing
		}
	}


	// moveNext()
	// If cursor is defined and not at back, moves cursor one step toward
	// back of this List, if cursor is defined and at back, cursor becomes
	// undefined, if cursor is undefined does nothing.
	void moveNext(){
		if(cursor != null && this.index() != this.length() - 1){
		cursor = cursor.next;
		cursorindex = cursorindex + 1;
		}
		else if(cursor != null && this.index() == this.length() - 1){
			cursor = null;
			cursorindex = -1;
	}
	else{
	}
	}
	
	
	// prepend()
	// Insert new element into this List. If List is non-empty,
	// insertion takes place before front element.
	void prepend(int data){
		Node N = new Node(data);
		if(this.length() == 0){
			front = back = N;
		}
		else{
			front.prev = N;
			N.next = front;
			front = N;
		}
		length++;
	} 
	
	// append()
	// Insert new element into this List. If List is non-empty,
	// insertion takes place after back element.
	void append(int data){
		Node N = new Node(data);
		if(this.length == 0){
			front = back = N;
		}
		else{
			back.next = N;
			N.prev = back;
			back = N;
		}
		length++;
	} 
	
	// insertBefore()
	// Insert new element before cursor.
	// Pre: length()>0, index()>=0
	void insertBefore(int data){
		if(this.length() <= 0){
			throw new RuntimeException("List Error: insertBefore() called on empty list");
		}
		if(this.index() < 0){
			throw new RuntimeException("List Error: insertBefore() called on undefined cursor");
		}
		Node N = new Node(data);
		if(cursor.prev == null){
			cursor.prev = N;
			N.next = cursor;
			front = N;
		}else{
			N.prev = cursor.prev;
			cursor.prev.next = N;
			N.next = cursor;
			cursor.prev = N;
		}
		length++;
		cursorindex++;
	}
	
	// insertAfter()
	// Inserts new element after cursor.
	// Pre: length()>0, index()>=0
	void insertAfter(int data){
		if(this.length() <= 0){
			throw new RuntimeException("List Error: insertAfter() called on empty list");
		}
		if(this.index() < 0){
			throw new RuntimeException("List Error: insertAfter() called on undefined cursor");
		}
		
		Node N = new Node(data);
		if(cursor.next == null){
			cursor.next = N;
			N.prev = cursor;
			back = N;
		}
		else{
			cursor.next.prev = N;
			N.next = cursor.next;
			N.prev = cursor;
			cursor.next = N;
		}
		length++;
	}
 
	// deleteFront()
	// Deletes the front element. 
	// Pre: length()>0
	void deleteFront(){
		if(this.length() <= 0){
			throw new RuntimeException("List Error: deleteFront() called on empty list");
		}
		front = front.next;
		front.prev = null;
		length--;
		}
	
	// deleteBack()
	// Deletes the back element. 
	// Pre: length()>0
	void deleteBack(){
		if(this.length() <= 0){
			throw new RuntimeException("List Error: deleteBack() called on empty list");
		}
		back = back.prev;
		back.next = null;
		length--;
		}
	
	
	//delete()
	// Deletes cursor element, making cursor undefined.
	// Pre: length()>0, index()>=0
	void delete(){
		if(this.length() <= 0){
			throw new RuntimeException("List Error: delete() called on empty list");
		}
		if(this.index() < 0){
			throw new RuntimeException("List Error: delete() called on undefined cursor");
		}
		if(cursor == front){
			deleteFront();
		}
		else if(cursor == back){
			deleteBack();
		}
		else{
			cursor.next.prev = cursor.prev;
			cursor.prev.next = cursor.next;
			length--;
			cursor = null;
			cursorindex = -1;
		}
	}
	
	
	
// Other methods

	// toString()
	// Overrides Object's toString method. Returns a String
	// representation of this List consisting of a space
	// separated sequence of integers, with front on left.
	public String toString(){
		StringBuffer sb = new StringBuffer();
		Node N = front;
		while(N!=null){
			sb.append(N.toString());
			sb.append(" ");
			N = N.next;
		}
		return new String(sb);
	}
	
	// copy()
	// Returns a new List representing the same integer sequence as this
	// List. The cursor in the new list is undefined, regardless of the
	// state of the cursor in this List. This List is unchanged.
	List copy(){
		List Q = new List();
		Node N = this.front;

		while( N!=null ){
			Q.append(N.data);
			N = N.next;
		}
		return Q;
	}
}
















	
